draft-ietf-jsonpath-base.md

```diff
@@ -1162,22 +1161,27 @@ logical-not-op      = "!"               ; logical NOT operator
 ~~~~
 
 A test expression
-either tests the existence of a node
-designated by an embedded query (see {{extest}}) or tests the
-result of a function expression (see {{fnex}}).
-In the latter case, if the function result type is declared as
-`LogicalType` (see {{typesys}}), it tests whether the result
-is `LogicalTrue`; if the function result type is declared as
-`NodesType`, it tests whether the result is non-empty.
-If the declared function result type is `ValueType`, its use in a
-test expression is not well-typed.
+either tests the existence of nodes
+designated by an embedded query or by a function expression of declared type
+`SingleNodeType` or `NodesType` (see {{extest}}) or tests the
+result of a function expression of declared type `LogicalType` (see
+{{fnex}}).
+In the latter case, the test expression tests whether the function
+result is `LogicalTrue`.
+If a declared function result type is any other than `SingleNodeType`, `NodesType`, or `LogicalType`,
+its use in a test expression is not well-typed.
 
 ~~~ abnf
-
 test-expr           = [logical-not-op S]
                       (filter-path /  ; path existence/non-existence
-                       function-expr) ; LogicalType or
-                                      ; NodesType
+                       logical-func /
+                       nodes-func /
+                       single-node-func)
+
+logical-func        = function-expr ; LogicalType
+nodes-func          = function-expr ; NodesType
+single-node-func    = function-expr ; SingleNodeType
```

Comment:
I realize this doesn't do much for automated ABNF parsers, but it helps readability, and I hope we might be able to find a way to distinguish these functions syntactically.

Reply:
Is there a way to have extensible ABNF?  Maybe we can do something like

```abnf
function-name         = value-func-name /
                        logical-func-name /
                        single-node-func-name /
                        nodes-func-name
value-func-name       = "length" /
                        "count" /
                        {???} ; a template that can be completed by an extension
logical-func-name     = "match"
                        "search" /
                        {???}
single-node-func-name = {???}
nodes-func-name       = {???}
```

I don't know if the `{???}` is possible in ABNF.

Reply:
The intention with the extensibility approach was to have a stable grammar (ABNF) plus semantic rules that checks validity of the constructed parse tree, where the latter make use of information from the registry.
I don't think we have technology to import registry information into ABNF (we are looking at that in CDDL, but even that won't be able to extend the grammar to express the semantic constraints, which include constraints on arguments).  Your additions look like they need a painful hand-written parser with a lexer that assigns different syntactical categories based on a symbol table lookup, like in a C compiler.  I'd like to stick with ABNF.

Reply:
Okay... I can revert this bit.  I'm not married to it.  It was just an idea.